### DISCUSSION TITLE
NO_FREE_ACTION_SLOTS

### DISCUSSION THREAD

**Discussion by Strykix** [YarostheLaunchpadder/MissionPlanner-to-Litchi#25] [Q&A]
Hello, 

i put the same parameters that you're describing into your readme, but i had an error message regarding the NO_FREE_ACTION_SLOTS, could you help ? =) 

thanks

**Comment by JoeKae** [comment]
Can you please supply the input file? Then i will check it later.

**Comment by Strykix** [comment]
Thanks JoeKae, here is the file
[test.zip](https://github.com/YarostheLaunchpadder/MissionPlanner-to-Litchi/files/10784241/test.zip)

on top of that it have done the conversion but the error was worrying me

**Reply by JoeKae** [reply]
I am super busy at the moment. I will have a look on the weekend.

**Reply by Strykix** [reply]
that's ok man, it is working when i upload it to litchi, but just i'm having this error message, there is maybe a reason behind let me know when you'll have time ;) 
on top of that when you're doing photogrametry what's your gimball angle ?

**Reply by JoeKae** [reply]
> on top of that when you're doing photogrametry what's your gimball angle ?

That depends on the terrain, the object and the distance. 
I did not have a lot of time trying things out with photogrammetry but when i did i went for these angles:

- low altitude
  - 10° - 15°
- medium altitude
  - close
    - 30° - 40°
  - far
    - 45° - 55°
- high altitude
  - 70° - 80°

Altitude is meant relative to the object.
Results were ok. Maybe i need a better camera for better models or i need better settings.

**Reply by JoeKae** [reply]
I did just have a look and could not reproduce the problem.
I do ignore the `Qt: Untested Windows version 10.0 detected!` message as it is just a warning.

Try these steps and tell me if you encounter the problem again.

# Steps

- checkout the main branch
- create a virtual environment with python 3.10
- install the requirements from `requirements.txt`
- run the file `run.py`
- select the test.waypoints file
- `test.waypoints.csv` is created without any prompts like `NO_FREE_ACTION_SLOTS`


# What is `NO_FREE_ACTION_SLOTS`

Litchi only allows 15 actions per waypoint. Mission Planner on the other hand has no limits like that. So it is possible to overwhelm Litchi with a Mission Planner flightplan. The Litchi file will still be created but there are actions dropped and the flightplan might not be as you expect it to be.
The file you posted did not have to many actions though. So i am not sure what might have gone wrong here. I attached the converted file for you so you can compare them.
[test.waypoints.csv](https://github.com/YarostheLaunchpadder/MissionPlanner-to-Litchi/files/10832422/test.waypoints.csv)

# Considering another problem

As you did use the [Waypoint trigger](https://github.com/YarostheLaunchpadder/MissionPlanner-to-Litchi/blob/main/docs/MP_cam_ctrl.md) workflow you could encounter an unwanted effect on your images. The altitude of 120m should be high enough though so the motion blur might not be noticable.
The problem was discussed in #22 and should be fixed in the other file. Here a 1 second long `Stay For` action is added after each photo capture. This can improve the image quality (motion blur reduction) and prevent the rotation images but has a longer flight time. If you encounter motion blur or images while the drone was rotating then you can give this file a shot.
[test.waypoints_stay_1s_after_capture.csv](https://github.com/YarostheLaunchpadder/MissionPlanner-to-Litchi/files/10832423/test.waypoints_stay_1s_after_capture.csv)
